File: Python/n_queens.py
def is_safe(board, row, col, N):
    # Check if there's a queen in the same column
    for i in range(row):
        if board[i][col] == 1:
            return False

    # Check upper-left diagonal
    for i, j in zip(range(row, -1, -1), range(col, -1, -1)):
        if board[i][j] == 1:
            return False

    # Check upper-right diagonal
    for i, j in zip(range(row, -1, -1), range(col, N)):
        if board[i][j] == 1:
            return False

    return True

def solve_n_queens(N):
    board = [[0 for _ in range(N)] for _ in range(N)]
    if solve_util(board, 0, N) is False:
        print("No solution exists")
    else:
        print_solution(board)

def solve_util(board, row, N):
    if row == N:
        return True

    for col in range(N):
        if is_safe(board, row, col, N):
            board[row][col] = 1

            if solve_util(board, row + 1, N):
                return True

            board[row][col] = 0  # Backtrack

    return False

def print_solution(board):
    for row in board:
        print(" ".join(["Q" if cell == 1 else "." for cell in row]))

N = 8  # Change N to solve for different board sizes
solve_n_queens(N)

File: JavaSolutions/NQueens.java
import java.util.*;

public class NQueens {
    public static void solveNQueens(int n) {
        char[][] board = new char[n][n];
        for (char[] row : board)
            Arrays.fill(row, '.'); 
        
        List<List<String>> solutions = new ArrayList<>();
        placeQueens(board, 0, solutions);
        
        for (List<String> sol : solutions) {
            for (String row : sol)
                System.out.println(row);
            System.out.println();
        }
    }

    private static void placeQueens(char[][] board, int row, List<List<String>> solutions) {
        int n = board.length;
        if (row == n) {
            solutions.add(construct(board));
            return;
        }

        for (int col = 0; col < n; col++) {
            if (isSafe(board, row, col)) {
                board[row][col] = 'Q';
                placeQueens(board, row + 1, solutions);
                board[row][col] = '.'; // backtrack
            }
        }
    }

    private static boolean isSafe(char[][] board, int row, int col) {
        int n = board.length;

        
        for (int i = 0; i < row; i++)
            if (board[i][col] == 'Q') return false;

    
        for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--)
            if (board[i][j] == 'Q') return false;


        for (int i = row - 1, j = col + 1; i >= 0 && j < n; i--, j++)
            if (board[i][j] == 'Q') return false;

        return true;
    }

    private static List<String> construct(char[][] board) {
        List<String> res = new ArrayList<>();
        for (char[] row : board)
            res.add(new String(row));
        return res;
    }

    public static void main(String[] args) {
        int n = 4; // Change this to test with different board sizes
        solveNQueens(n);
    }
}
